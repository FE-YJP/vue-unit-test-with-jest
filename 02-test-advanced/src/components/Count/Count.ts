import Vue from 'vue'

export default Vue.extend({
  data: () => ({
  }),

  computed: {
    /**
     * 当前vue版本下不建议使用mapGetters/mapState/mapActions/mapMutations函数
     * 因为：1.ts不能识别其类型，2.对测试不友好
     * vue3.x开始对ts支持可能会更好
     */
    count (): number {
      return this.$store.state.count.count
    },

    evenOrOdd (): 'even'|'odd' {
      return this.$store.getters.evenOrOdd
    }
  },

  methods: {
    plus () {
      this.$store.dispatch('increment')
    },

    minus () {
      this.$store.dispatch('decrement')
    }
  }
})
