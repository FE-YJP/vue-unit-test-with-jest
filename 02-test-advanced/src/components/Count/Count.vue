<template>
  <div class="ml-4">
    <div> store 测试：</div>
    <div class="mt-2">
      <v-btn small @click="minus" color="primary" class="mr-2">-</v-btn>
       {{ count }}
       <v-btn small @click="plus" color="primary" class="ml-2">+</v-btn>

       <v-chip class="ma-2"> {{ evenOrOdd }}</v-chip>
    </div>
  </div>
</template>

<script lang="ts" src="./Count.ts"></script>
