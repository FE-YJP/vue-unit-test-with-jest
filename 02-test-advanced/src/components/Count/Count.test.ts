/**
 * 测试思路：
 * （本例是以$store为例，测试时遇到$router/$route的mock也是一样的）
 * 1. 创建mount/shallowMount的挂载项options
 * 2. mock掉$store（包括其内部的state/getters/dispatch/commit）
 * 3. 创建wrapper时记得挂载options
 * 4. 测试需测试项
 */

import { shallowMount } from '@vue/test-utils'
import Count from './Count.vue'

describe('components.Count', () => {
  let wrapper: any
  let options: any
  const mockDispatch = jest.fn()

  beforeEach(() => {
    options = {
      // https://vue-test-utils.vuejs.org/zh/api/options.html#%E6%8C%82%E8%BD%BD%E9%80%89%E9%A1%B9
      propsData: {}, // 组件无props，存在的情况下传递即可
      stubs: {}, // 子组件无需存根
      mocks: {
        /**
         * mocks 中最长出现的是，$store/$route/$router
         * 用法类似，组件内访问store/route/router下的数据/方法时
         * 在此处mock一个对应的即可
         *
         * 注：我们只需要mock即可，具体的内部逻辑已经在store测试内测试过了
         *     这也符合单元测试的基本理念，只测试目标内容，以外的全部使用mock思想处理
         */

        $store: {
          state: {
            count: {
              count: 0 //mocak data，对应 this.$store.state.count.count，以下类似
            }
          },

          getters: {
            evenOrOdd: 'even' // mock data，此处因为也可以写作函数形式() => 'even'
          },

          dispatch: mockDispatch // mock fn
        }
      }

      /**
       * 可挂载项比较多，此处仅列出了常用的propsData/stubs/mocks
       * 上面的官网列出了其他选项
       * 这里要说明的是，官网的最后有一个[其他选项](https://vue-test-utils.vuejs.org/zh/api/options.html#%E5%85%B6%E5%AE%83%E9%80%89%E9%A1%B9)，这个就厉害了
       * 原理是使用了vue的extends API
       * 实际使用的时候，这里还可以挂载data/computed/methods来覆盖组件内原有的
       * 当然，也可以挂载任何其他需要的，都会以扩展的形式挂到wrapper上
       */
    }

    /**
     * 本测试用例比较简单，每次生成的wrapper都是相同的，
     * 所以可以将每个每个testcase内的wrapper删掉，集中写在此处
     *
     * 实际的项目中，不建议写在此处，因为即便有一处的options时需要变更的话，都不能集中写
     * 而且，如果出错的话，分开写更容易查错
     */

    // wrapper = shallowMount(Count, options)
  })

  describe('snapshot', () => {
    it('snapshot测试', () => {
      wrapper = shallowMount(Count, options)
      expect(wrapper.vm.$el).toMatchSnapshot();
    })
  });

  describe('methods.plus', () => {
    it('dispatch函数执行，且参数为increment', () => {
      wrapper = shallowMount(Count, options)
      wrapper.vm.plus()
      expect(mockDispatch).toBeCalledWith('increment')
    })
  });

  describe('methods.minus', () => {
    it('dispatch函数执行，且参数为decrement', () => {
      wrapper = shallowMount(Count, options)
      wrapper.vm.minus()
      expect(mockDispatch).toBeCalledWith('decrement')
    })
  });

  /**
   * computed 内的要不要测试？
   * 1. 只是获取数据，没有任何业务逻辑，没必要测试
   * 2. 最终反映到了页面上，snapshot测试就已经包含了
   */
})
