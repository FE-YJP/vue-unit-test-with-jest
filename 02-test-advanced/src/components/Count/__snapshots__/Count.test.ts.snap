// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components.Count snapshot snapshot测试 1`] = `
<div
  class="ml-4"
>
  <div>
     store 测试：
  </div>
   
  <div
    class="mt-2"
  >
    <v-btn-stub
      activeclass=""
      class="mr-2"
      color="primary"
      small="true"
      tag="button"
      type="button"
    >
      -
    </v-btn-stub>
    
     0
     
    <v-btn-stub
      activeclass=""
      class="ml-2"
      color="primary"
      small="true"
      tag="button"
      type="button"
    >
      +
    </v-btn-stub>
     
    <v-chip-stub
      active="true"
      activeclass=""
      class="ma-2"
      closeicon="$delete"
      filtericon="$complete"
      tag="span"
    >
       even
    </v-chip-stub>
  </div>
</div>
`;
